### ISSUE TITLE
Fix  proguard enabled build (Android )

### ISSUE THREAD

**Issue by mb12** [mapbox/mapbox-gl-native#2130]
Proguard enabled build fails with tons of warnings like the following.

```
Warning: org.apache.commons.beanutils.BeanMap$3: can't find referenced class org.apache.commons.collections.Transformer
Warning: org.apache.commons.beanutils.BeanMap$4: can't find referenced class org.apache.commons.collections.Transformer
Warning: org.apache.commons.beanutils.BeanMap$5: can't find referenced class org.apache.commons.collections.Transformer
Warning: org.apache.commons.beanutils.BeanMap$6: can't find referenced class org.apache.commons.collections.Transformer
Warning: org.apache.commons.beanutils.BeanMap$7: can't find referenced class org.apache.commons.collections.Transformer

```

**Comment by tobrunvannuland** [comment]
To support proguard in an AAR project following should be added to the build.gradle

    release {
        consumerProguardFiles 'proguard-android.txt'
    }

This file will be packaged in the AAR and will be automatically applied in the main application project.

**Comment by ljbade** [comment]
@mb12 Does @tobrunvannuland suggestion fix the problem?

**Comment by mb12** [comment]
@ljbade  No, it doesn't fix the problem. Proguard is already setup in the sample test app. Its just disabled. Changing minifyEnabled from false to true in build.gradle makes it run. However the build fails with tons of warnings. Proguard optimizes and minimizes apk size at bytecode level.  So  new rules would need to be added for the Mapbox SDK. It should easily knock  off a couple of megs from the final apk.  Once the proguard configuration file for Mapbox SDK is ready, it could be bundled into AAR project as @tobrunvannuland  suggest (I haven't tried it though).

**Comment by tobrunvannuland** [comment]
Will do some tests in next couple days, will produce some proguard configs along the way.

**Comment by mb12** [comment]
Thanks @tobrunvannuland , it will be very useful for Android. The APK file size limit is only 50MB for Android (half that of iOS). Since we build a fat apk that contains native libs of all supported platforms, the proguard optimization to reduce apk size will be very helpful.

**Comment by tobrunvannuland** [comment]
I have this setup working for the SNAPSHOT, while having the test app use following proguard config:

> -keep class com.mapbox.mapboxgl.annotations.** { *; }
> -keep class com.mapbox.mapboxgl.geometry.** { *; }
> -keep class com.mapbox.mapboxgl.http.** { *; }
> -keep class com.mapbox.mapboxgl.views.MapView**
> -keep class com.mapbox.mapboxgl.views.NativeMapView**
> -keepclassmembers class com.mapbox.mapboxgl.views.NativeMapView** {*;}
> -dontwarn okio.** 

Things to do:
 - Move config file to library project 
 - Test release build while building from source 
 - Optimise configuration further

Current result: 
 - apk size size was reduced from 13.8 MB to 13.2 MB 

Reason behind the limited result:
- Proguard only works with java bytecode and the amount of java code is currently limited. 

FYI: I took a look at the unzipped version of the apk and I noticed following:
 - total size is 37.5 MB 
 - lib folder size with native libs is 34.4 MB. 

This means that >90% of an obfuscated apk size is used up by the native libs

**Comment by tobrunvannuland** [comment]
Implemented in [PR](https://github.com/mapbox/mapbox-gl-native/pull/2206)

**Comment by ljbade** [comment]
@tobrunvannuland Yeah we still working on reducing the native libs size, mostly by replace bits of dependencies with Java implementations: https://github.com/mapbox/mapbox-gl-native/issues/2037

**Comment by bleege** [comment]
I'll look at merging #2206.  Thanks @tobrunvannuland!

**Comment by bleege** [comment]
Tested and merged #2206.  Thanks again @tobrunvannuland!
